### DISCUSSION TITLE
Android Tinycam audio?

### DISCUSSION THREAD

**Discussion by aleqcz** [thirtythreeforty/neolink#361] [General]
Hi,
is there anybody using Neolink with Tinycam? Does the audio work for you? Whenever I press the audio button on top, it turns organge (activates), but the circle animater next to it and then it display "Failed to connect audio (timeout, check hostname and port)".
However VLC works...
Is it a tinycam settings or something?

Thanks for any tips,
A.

**Comment by QuantumEntangledAndy** [comment]
I'm not familiar with tinycam. Is this button to activate sound or to activate the speaker (I assume via onvif). There is no seperate port/connection for audio it's all sent though the same stream so your error message is odd unless tinycam is trying to eatabilsh a speaking connection.

**Reply by QuantumEntangledAndy** [reply]
Perhaps you could send a screen cap or screen shot

**Comment by aleqcz** [comment]
Ok, it doesn't work today at all. It's throwing those segfaults which I used to get a year ago and which were accounted to docker. I was under the impression the issue is fixed but I'm getting it over and over today:
`neolink_1_984855be6cbc | [2023-06-16T09:04:03Z INFO  neolink] Neolink 0.4.0 (unknown commit) release
neolink_1_984855be6cbc | [2023-06-16T09:04:03Z WARN  neolink::rtsp] Without a server certificate, usernames and passwords will be exchanged in plaintext!
neolink_1_984855be6cbc | [2023-06-16T09:04:03Z WARN  neolink::rtsp] The format config option of the camera has been removed in favour of auto detection.
neolink_1_984855be6cbc | [2023-06-16T09:04:03Z INFO  neolink::rtsp] e1: Connecting to camera at Address: 192.168.1.82:9000
neolink_1_984855be6cbc | [2023-06-16T09:04:03Z INFO  neolink::rtsp] e1: Logging in
neolink_1_984855be6cbc | [2023-06-16T09:04:03Z INFO  neolink::rtsp] e1: Connecting to camera at Address: 192.168.1.82:9000
neolink_1_984855be6cbc | [2023-06-16T09:04:03Z INFO  neolink::rtsp] e1: Logging in
neolink_1_984855be6cbc | [2023-06-16T09:04:03Z INFO  neolink::rtsp] e1: Connected and logged in
neolink_1_984855be6cbc | [2023-06-16T09:04:03Z INFO  neolink::rtsp] e1: Starting video stream Sub Stream (Fluent)
neolink_1_984855be6cbc | [2023-06-16T09:04:03Z INFO  neolink::rtsp] e1: Connected and logged in
neolink_1_984855be6cbc | [2023-06-16T09:04:03Z INFO  neolink::rtsp] e1: Camera time is already set: 2023-06-16 11:04:04 +1
neolink_1_984855be6cbc | [2023-06-16T09:04:03Z INFO  neolink::rtsp] e1: Camera reports firmware version v3.0.0.716_21112404
neolink_1_984855be6cbc | [2023-06-16T09:04:03Z INFO  neolink::rtsp] e1: Starting video stream Main Stream (Clear)
neolink_1_984855be6cbc |
neolink_1_984855be6cbc | (neolink:8): GLib-GObject-CRITICAL **: 09:04:37.559: g_object_new_is_valid_property: property 'name' for type 'GstRtpL16Pay' cannot be set twice
neolink_1_984855be6cbc |
neolink_1_984855be6cbc | (neolink:8): GStreamer-CRITICAL **: 09:04:38.006: gst_mini_object_lock: assertion 'GST_MINI_OBJECT_IS_LOCKABLE (object)' failed
neolink_1_984855be6cbc | Segmentation fault (core dumped)
neolink_1_984855be6cbc | [2023-06-16T09:04:39Z WARN  neolink::rtsp] Without a server certificate, usernames and passwords will be exchanged in plaintext!
neolink_1_984855be6cbc | [2023-06-16T09:04:39Z WARN  neolink::rtsp] The format config option of the camera has been removed in favour of auto detection.
neolink_1_984855be6cbc | [2023-06-16T09:04:39Z INFO  neolink::rtsp] e1: Connecting to camera at Address: 192.168.1.82:9000
neolink_1_984855be6cbc | [2023-06-16T09:04:39Z INFO  neolink::rtsp] e1: Logging in
neolink_1_984855be6cbc | [2023-06-16T09:04:39Z INFO  neolink::rtsp] e1: Connecting to camera at Address: 192.168.1.82:9000
neolink_1_984855be6cbc | [2023-06-16T09:04:39Z INFO  neolink::rtsp] e1: Logging in
neolink_1_984855be6cbc | [2023-06-16T09:04:39Z INFO  neolink::rtsp] e1: Connected and logged in
neolink_1_984855be6cbc | [2023-06-16T09:04:39Z INFO  neolink::rtsp] e1: Starting video stream Sub Stream (Fluent)
neolink_1_984855be6cbc | [2023-06-16T09:04:39Z INFO  neolink::rtsp] e1: Connected and logged in
neolink_1_984855be6cbc | [2023-06-16T09:04:39Z INFO  neolink::rtsp] e1: Camera time is already set: 2023-06-16 11:04:40 +1
neolink_1_984855be6cbc | [2023-06-16T09:04:39Z INFO  neolink::rtsp] e1: Camera reports firmware version v3.0.0.716_21112404
neolink_1_984855be6cbc | [2023-06-16T09:04:39Z INFO  neolink::rtsp] e1: Starting video stream Main Stream (Clear)
neolink_1_984855be6cbc |
neolink_1_984855be6cbc | (neolink:7): GLib-GObject-CRITICAL **: 09:04:43.122: g_object_new_is_valid_property: property 'name' for type 'GstRtpL16Pay' cannot be set twice
neolink_1_984855be6cbc |
neolink_1_984855be6cbc | (neolink:7): GStreamer-CRITICAL **: 09:04:43.847: gst_mini_object_lock: assertion 'GST_MINI_OBJECT_IS_LOCKABLE (object)' failed
neolink_1_984855be6cbc | Segmentation fault (core dumped)
neolink_neolink_1_984855be6cbc exited with code 139
neolink_1_984855be6cbc | [2023-06-16T09:04:44Z WARN  neolink::rtsp] Without a server certificate, usernames and passwords will be exchanged in plaintext!
neolink_1_984855be6cbc | [2023-06-16T09:04:44Z WARN  neolink::rtsp] The format config option of the camera has been removed in favour of auto detection.
neolink_1_984855be6cbc | [2023-06-16T09:04:44Z INFO  neolink::rtsp] e1: Connecting to camera at Address: 192.168.1.82:9000
neolink_1_984855be6cbc | [2023-06-16T09:04:44Z INFO  neolink::rtsp] e1: Logging in
neolink_1_984855be6cbc | [2023-06-16T09:04:44Z INFO  neolink::rtsp] e1: Connecting to camera at Address: 192.168.1.82:9000
neolink_1_984855be6cbc | [2023-06-16T09:04:44Z INFO  neolink::rtsp] e1: Logging in
neolink_1_984855be6cbc | [2023-06-16T09:04:44Z INFO  neolink::rtsp] e1: Connected and logged in
neolink_1_984855be6cbc | [2023-06-16T09:04:44Z INFO  neolink::rtsp] e1: Starting video stream Sub Stream (Fluent)
neolink_1_984855be6cbc | [2023-06-16T09:04:45Z INFO  neolink::rtsp] e1: Connected and logged in
neolink_1_984855be6cbc | [2023-06-16T09:04:45Z INFO  neolink::rtsp] e1: Camera time is already set: 2023-06-16 11:04:45 +1
neolink_1_984855be6cbc | [2023-06-16T09:04:45Z INFO  neolink::rtsp] e1: Camera reports firmware version v3.0.0.716_21112404
neolink_1_984855be6cbc | [2023-06-16T09:04:45Z INFO  neolink::rtsp] e1: Starting video stream Main Stream (Clear)
neolink_1_984855be6cbc |
neolink_1_984855be6cbc | (neolink:7): GLib-GObject-CRITICAL **: 09:04:48.988: g_object_new_is_valid_property: property 'name' for type 'GstRtpL16Pay' cannot be set twice
neolink_1_984855be6cbc |
neolink_1_984855be6cbc | (neolink:7): GStreamer-CRITICAL **: 09:04:49.691: gst_mini_object_lock: assertion 'GST_MINI_OBJECT_IS_LOCKABLE (object)' failed
neolink_1_984855be6cbc | Segmentation fault (core dumped)
neolink_neolink_1_984855be6cbc exited with code 139
`

Sorry for hijacking the topic, should I rather create a new one?

Many thanks for previous replies

**Reply by QuantumEntangledAndy** [reply]
Which image are you using? Ah. I just noticed we are on the thirtythreeforty/neolink. Please check out and use my fork instead this one is mostly unmaintained. https://github.com/QuantumEntangledAndy/neolink/

**Comment by aleqcz** [comment]
Oh yes, I'm on thirtythreeforty... I'll try quantumentangledandy/neolink, thanks for tip

My docker-compose.yml before the change:

```
version: '3.3'
services:
    neolink:
        ports:
            - '8554:8554'
        restart: unless-stopped
        volumes:
            - './config.toml:/etc/neolink.toml'
        image: thirtythreeforty/neolink
```

**Comment by aleqcz** [comment]
Nice, the picture is back up, thanks! I noticed the original docker image was using 0.4.0...

back to the original topic - audio. Thanks to the latest version I managed to glimpse an error in TinyCam about the audio format:
![image](https://github.com/thirtythreeforty/neolink/assets/1750538/2101e472-efb1-4f58-aeb5-6673589598c1)
Translation: Unsupported codec. Switch camera codec to Linear PCM, G.711 A-Law (PCMA), G.711 u-law (PCMU), AAC, AMR-NB, G.726 (32kbps), Opus.

This provides the reason, I'll try to find out more on this topic - like which codec is used and if TinyCam can be reconfigured or something.
Strangely, if I connect to E1 Pro camera directly using "Reolink C1" preset, it kind of works (audio 100 %, video mostly until it freezes and requires manual reconnect). This leads me to thinking the camera's audio code is supported by TinyCam. Is Neolink reencoding the audio?

**Reply by QuantumEntangledAndy** [reply]
Yeah we re-encode the aac into linear pcm because it's supported by more players. It should be linear pcm 16bit big endian which is pretty much just raw audio. Anything should play it. I wonder why the player is having trouble. Maybe you could run the stream through ffprobe to confirm the format.

**Reply by aleqcz** [reply]
ffprobe output confirms:
```
$ ffprobe -rtsp_transport tcp -hide_banner -show_streams "rtsp://admin:XXX@
192.168.1.1:8554/e1"
[rtsp @ 0x55c9f25a7b80] decoding for stream 0 failed
Input #0, rtsp, from 'rtsp://admin:XXX@192.168.1.1:8554/e1':
  Metadata:
    title           : Session streamed with GStreamer
    comment         : rtsp-server
  Duration: N/A, start: 0.000000, bitrate: 512 kb/s
    Stream #0:0: Video: h264 (High), yuv420p(progressive), 2560x1440, 90k tbr, 90k tbn, 180k tbc
    Stream #0:1: Audio: pcm_s16be, 16000 Hz, 2 channels, s16, 512 kb/s
[STREAM]
index=0
codec_name=h264
codec_long_name=H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
profile=High
codec_type=video
codec_time_base=0/2
codec_tag_string=[0][0][0][0]
codec_tag=0x0000
width=2560
height=1440
coded_width=2560
coded_height=1440
has_b_frames=0
sample_aspect_ratio=N/A
display_aspect_ratio=N/A
pix_fmt=yuv420p
level=51
color_range=unknown
color_space=unknown
color_transfer=unknown
color_primaries=unknown
chroma_location=left
field_order=progressive
timecode=N/A
refs=1
is_avc=false
nal_length_size=0
id=N/A
r_frame_rate=90000/1
avg_frame_rate=0/0
time_base=1/90000
start_pts=N/A
start_time=N/A
duration_ts=N/A
duration=N/A
bit_rate=N/A
max_bit_rate=N/A
bits_per_raw_sample=8
nb_frames=N/A
nb_read_frames=N/A
nb_read_packets=N/A
DISPOSITION:default=0
DISPOSITION:dub=0
DISPOSITION:original=0
DISPOSITION:comment=0
DISPOSITION:lyrics=0
DISPOSITION:karaoke=0
DISPOSITION:forced=0
DISPOSITION:hearing_impaired=0
DISPOSITION:visual_impaired=0
DISPOSITION:clean_effects=0
DISPOSITION:attached_pic=0
DISPOSITION:timed_thumbnails=0
[/STREAM]
[STREAM]
index=1
codec_name=pcm_s16be
codec_long_name=PCM signed 16-bit big-endian
profile=unknown
codec_type=audio
codec_time_base=1/16000
codec_tag_string=[0][0][0][0]
codec_tag=0x0000
sample_fmt=s16
sample_rate=16000
channels=2
channel_layout=unknown
bits_per_sample=16
id=N/A
r_frame_rate=0/0
avg_frame_rate=0/0
time_base=1/16000
start_pts=0
start_time=0.000000
duration_ts=N/A
duration=N/A
bit_rate=512000
max_bit_rate=N/A
bits_per_raw_sample=N/A
nb_frames=N/A
nb_read_frames=N/A
nb_read_packets=N/A
DISPOSITION:default=0
DISPOSITION:dub=0
DISPOSITION:original=0
DISPOSITION:comment=0
DISPOSITION:lyrics=0
DISPOSITION:karaoke=0
DISPOSITION:forced=0
DISPOSITION:hearing_impaired=0
DISPOSITION:visual_impaired=0
DISPOSITION:clean_effects=0
DISPOSITION:attached_pic=0
DISPOSITION:timed_thumbnails=0
[/STREAM]
```

**Comment by aleqcz** [comment]
Is there any configuration option to change the audio codec (I'd like to try and see if it helps).

Or anybody successfully used this with TinyCam somehow? It's more for discussion with them rather than this project, right?

**Reply by aleqcz** [reply]
Can I open a feature request somehow, any chance for this or is it not likely to happen?
